### ISSUE TITLE
Multiple tycheck errors in VST

### ISSUE THREAD

**Issue by ildyria** [PrincetonUniversity/VST#133]
I tried to compile and verify the following function:

```c
u8 test_simpl(const i64 n)
{
  i64 b,c;
  u8 d;
  b = n * 2;
  c = 2 * n;
  
  b = b + 4;
  b = (-4) + b;
  
  c = c - 4;
  c = 4 - c;
  
  b = b >> 8;
  c = c << 8;
  
  d = c & 0xff;
  d = d & b;
  return d;
}
```

With this specification:

```Coq
Definition test_simpl_spec :=
 DECLARE _test_simpl
  WITH n: Z
  PRE [ _n OF tlong ]
        PROP  (0 <= n < Int.max_signed)
        LOCAL (temp _n (Vlong (Int64.repr n)))
        SEP   ()
  POST [ tuchar ]
        PROP ()
        LOCAL(temp ret_temp (Vint (Int.repr 0)))
SEP ().
```

## Compilation with gcc and compcert

The test.c file can be compiled as follow:

```bash
gcc test.c -Wall -Wextra -o test  
```
or with Compcert (2.7.2)
```bash
ccomp test.c -o test
```
or interpreted with Compcert (2.7.2)
```bash
ccomp -interp test.c
```

In all these cases, no error or warning will be returned which leads to the thought that `test.c` is a valide C11 file.

## VST typechecking

In `test.c` I test the different operations (  `+`, `-`, `*`, `>>` and `<<`) with either side of the operand being a constant (thus of type `Int`).
I also test assignment to a uchar which does not typecast.

The `.v` file contains the `clightgen` output of `test.c` and thus can directly be opened in coqide.

More details: https://gist.github.com/ildyria/8604984298dfcc9234e776be4197181b

## TL;DR:

`long long = long long op int` typecheck but
`long long = int op long long` does not
`unsigned char = long long` does not typecheck at all even if the reduction `mod 255` is available.

**Comment by andrew-appel** [comment]
I'll work on this after I finish the port to CompCert 3.1, since these involve overlapping changes to the typechecker.

**Comment by andrew-appel** [comment]
This will be fixed in the next commit.    The test case is progs/cast_test.c, progs/verif_cast_test.v.
Other broken casts (not tested in this example) are also fixed.
